### PR TITLE
LogBroadcaster should not be sending logs for re-orged chains

### PR DESCRIPTION
**Pull request by PiotrTrzpil**
If (in the moment of sending the log) the log's block is no longer in the longest chain, it is skipped and not sent to any listeners.

The test changes were needed as they frequently used disconnected chain fragments.

### CODE REVIEW

File: core/services/log/broadcaster_test.go [hunk 1]
```diff
@@ -1201,32 +1197,39 @@ func TestBroadcaster_ProcessesLogsFromReorgsAndMissedHead(t *testing.T) {
 		log1R       = cltest.RawNewRoundLog(t, addr, blockHash1R, 1, 0, false)
 		log2R       = cltest.RawNewRoundLog(t, addr, blockHash2R, 2, 0, false)
 
-		head0 = models.Head{Hash: blockHash0, Number: 0}
-		// head1 - missing
```
Comment (PiotrTrzpil) [inline core/services/log/broadcaster_test.go:1205]: not sure what scenario this was covering - we should never have a missing head in the chain normally

File: core/services/log/registrations.go [hunk 2]
```diff
@@ -210,6 +210,14 @@ func (r *registrations) sendLog(log types.Log, orm ORM, latestHead models.Head,
 			continue
 		}
 
+		if !latestHead.IsInChain(log.BlockHash) {
+			logger.Tracew("Skipping because not in canonical ",
```
Comment (samsondav) [inline core/services/log/registrations.go:214]: should it say "not in canonical longest chain" ?

**Review by samsondav — APPROVED** [review]

File: core/services/log/registrations.go (hunk 2 above)
Comment (PiotrTrzpil) [reply core/services/log/registrations.go]: removed this part for now

File: core/services/headtracker/head_tracker.go [hunk 3]
```diff
@@ -133,10 +132,6 @@ func (ht *HeadTracker) HighestSeenHeadFromDB() (*models.Head, error) {
 	return ht.headSaver.HighestSeenHeadFromDB()
 }
 
-func (ht *HeadTracker) Chain(ctx context.Context, hash common.Hash, depth uint) (models.Head, error) {
```
[the remark this answers is not in the document]
Comment (PiotrTrzpil) [reply core/services/headtracker/head_tracker.go:136]: yup, wasn't used

File: core/services/log/pool.go [hunk 4]
```diff
@@ -1,50 +1,99 @@
...
-		logger.Tracew(fmt.Sprintf("LogBroadcaster: Will delete %v older logs", len(pool.allLogs)-len(logsToKeep)),
-			"latestBlockNum", latestBlockNum,
-			"highestNumConfirmations", highestNumConfirmations,
-			"remainingLogsCount", len(logsToKeep),
-			"finalityDepth", finalityDepth,
-		)
-		pool.allLogs = logsToKeep
 	}
+
+	for {
```
[the remark this answers is not in the document]
Comment (PiotrTrzpil) [reply core/services/log/pool.go:60]: added - thanks!

File: core/services/log/pool.go [hunk 5]
```diff
@@ -1,50 +1,99 @@
...
+			delete(pool.logsByBlockHash, hash)
+		}
+		delete(pool.hashesByBlockNumbers, blockNum)
+	}
+
 	return logsToReturn
 }
+
+func (pool *logPool) removeLog(log types.Log) {
+	// deleting all logs for this log's block hash
```
[the remark this answers is not in the document]
Comment (PiotrTrzpil) [reply core/services/log/pool.go:82]: it was a bit optional, as they would get deleted eventually - but yeah i added for consistency

File: core/internal/cltest/cltest.go [hunk 6]
```diff
@@ -2039,6 +2012,59 @@ func SimulateIncomingHeads(t *testing.T, args SimulateIncomingHeadsArgs) (func()
...
+	hashes := make([]common.Hash, 0)
+	heads := make(map[int64]*models.Head)
+	for i := int64(0); i < int64(numHashes); i++ {
+		hash := NewHash()
+		hashes = append(hashes, hash)
+
+		heads[i] = &models.Head{Hash: hash, Number: i}
+		if i > 0 {
+			parent, exists := heads[i-1]
+			if exists {
```
[the remark this answers is not in the document]
Comment (PiotrTrzpil) [reply core/internal/cltest/cltest.go:2048]: yes indeed

File: core/services/log/pool.go [hunk 7]
```diff
@@ -1,50 +1,111 @@
...
 		keptLogsDepth = highestNumConfirmations
 	}
-	// deleting old logs that will never be sent for any listener anymore
-	if len(pool.allLogs) > 0 {
-		for _, log := range pool.allLogs {
-			if int64(log.BlockNumber) >= int64(latestBlockNum)-int64(keptLogsDepth) {
-				logsToKeep = append(logsToKeep, log)
-			}
+
+	keptDepth := int64(latestBlockNum) - int64(keptLogsDepth)
```
Comment (connorwstein) [inline core/services/log/pool.go:42]: seems a little weird to compute this inside this func when its only used outside - can just compute it in deleteOldLogs

File: core/services/log/pool.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (connorwstein) [reply core/services/log/pool.go:85]: this way the data structures always remain consistent, seems like a nice property

File: core/services/log/pool.go (hunk 7 above)
Comment (PiotrTrzpil) [reply core/services/log/pool.go]: good point - moving this out simplified it quite a bit

File: core/services/log/broadcaster_test.go [hunk 8]
```diff
@@ -1201,32 +1197,39 @@ func TestBroadcaster_ProcessesLogsFromReorgsAndMissedHead(t *testing.T) {
...
 			head1R, log1Removed, log2Removed, log1R,
 			head2R, log2R,
 			head3R,
+			head4R,
 		}
 
-		expected = []types.Log{log0, log1, log2, log1R, log2R}
+		expectedA = []types.Log{log0, log1, log2, log1R, log2R}
+
+		// listenerB needs 3 confirmations, so log2 is not sent to it as it was reorged on head1R
```
[the remark this answers is not in the document]
Comment (PiotrTrzpil) [reply core/services/log/broadcaster_test.go:1338]: ok i now added a double-reorg. But 'reorged at the number of confirmations needed' was already there i think? Head 3 and then reorged Head 1, 2, 3 (one of the subscribers has 3 required confirmations). Or did you mean something slightly different?

[the remark this answers is not in the document]
Comment (connorwstein) [reply core/services/log/broadcaster_test.go]: yeah slightly different - I meant like the fork starts at the block which would have confirmed the log. However I think that case is actually covered by the 1 conf subscriber expected output so all good

[the remark this answers is not in the document]
Comment (connorwstein) [reply core/services/log/broadcaster_test.go]: this is a nice test good work

**Review by connorwstein — APPROVED** [review]

File: core/services/log/registrations.go [hunk 9]
```diff
@@ -164,12 +154,20 @@ func (r *registrations) isAddressRegistered(address common.Address) bool {
...
+	}
+
 	for _, log := range logs {
-		r.sendLog(log, orm, latestHead, &updates)
+		r.sendLog(log, latestHead, broadcastsExisting)
 	}
-	applyListenerInfoUpdates(updates, latestHead)
+}
+
+type Key struct {
```
Comment (connorwstein) [inline core/services/log/registrations.go:169]: leftover?

Comment (PiotrTrzpil) [reply core/services/log/registrations.go]: ah yes, removed now

**Review by connorwstein — APPROVED** [review]

**Pull request merged**